File: Gstreamer-Applications/gst_encode_decode/mp4_decode_playback.c
/*
* Copyright (c) 2020 Qualcomm Innovation Center, Inc.  All Rights Reserved.
*
* SPDX-License-Identifier: BSD-3-Clause-Clear
*/

/*
 * mp4_decode_playback.c:
 *
 * A sample app based on gstreamer
 * The purpose is helping users to learn how to implement the
 * mp4 file decoding to playback on weston with gstreamer on the
 * Qualcomm platform through this sample app.
 */

#include <gst/gst.h>
#include <glib.h>


static gboolean bus_callback(GstBus *bus, GstMessage *msg, gpointer data)
{
    GMainLoop *loop = (GMainLoop *)data;

    int type = GST_MESSAGE_TYPE(msg);
    
    if (type == GST_MESSAGE_EOS) {
        g_main_loop_quit (loop);
    } else if (type == GST_MESSAGE_ERROR) {
        gchar  *debug;
        GError *error;
        gst_message_parse_error(msg, &error, &debug);
        g_free(debug);
        g_printerr("Error: %s\n", error->message);
        g_error_free(error);
        g_main_loop_quit(loop);
    }

    return TRUE;
}


static void on_pad_added(GstElement *bus, GstPad *pad, gpointer data)
{
    GstElement *srcpad = (GstElement *)data;
    GstPad *sinkpad;
  
    sinkpad = gst_element_get_static_pad(srcpad, "sink");
    gst_pad_link(pad, sinkpad);
    gst_object_unref(sinkpad);
}



int main(int argc, char *argv[])
{
    GMainLoop *loop;
    GstElement *pipeline, *source, *demuxer, *decoder, *videoconv, *sink;
    GstBus *bus;
    guint bus_id;

    gst_init(&argc, &argv);

    loop = g_main_loop_new(NULL, FALSE);

    if (argc != 2) {
        g_printerr("Usage: %s <mp4 filename>\n", argv[0]);
        return -1;
    }

    pipeline    = gst_pipeline_new("video-player");
    source      = gst_element_factory_make("filesrc",      "file-source");
    demuxer     = gst_element_factory_make("qtdemux",      "qt-demuxer");
    decoder     = gst_element_factory_make("avdec_h264",   "h264-decoder");
    videoconv   = gst_element_factory_make("videoconvert", "converter");
    sink        = gst_element_factory_make("waylandsink",  "video-output");

    if (!pipeline || !source || !demuxer || !decoder || !videoconv || !sink) {
        g_printerr("Create element failed.\n");
        return -1;
    }

    g_object_set(G_OBJECT(source), "location", argv[1], NULL);

    g_object_set(G_OBJECT(sink), "x", 100, "y", 100, "width", 1280, "height", 720, NULL);

    // Or we set the output fullscreen to the sink element
    // g_object_set(G_OBJECT(sink), "fullscreen", 1, NULL);

    bus = gst_pipeline_get_bus(GST_PIPELINE(pipeline));
    bus_id = gst_bus_add_watch(bus, bus_callback, loop);
    gst_object_unref(bus);

    gst_bin_add_many(GST_BIN(pipeline), source, demuxer, decoder, videoconv, sink, NULL);

    gst_element_link(source, demuxer);
    gst_element_link_many(decoder, videoconv, sink, NULL);
    g_signal_connect(demuxer, "pad-added", G_CALLBACK(on_pad_added), decoder);

    gst_element_set_state(pipeline, GST_STATE_PLAYING);
    g_print("Start\n");
    g_main_loop_run(loop);
    g_print("Stop\n");
    gst_element_set_state(pipeline, GST_STATE_NULL);

    gst_object_unref(GST_OBJECT(pipeline));
    g_source_remove(bus_id);
    g_main_loop_unref(loop);

    gst_deinit();

    return 0;
}

File: RB5/linux_kernel_4_x/GStreamer-apps/python/gst_encode_decode/mp4_decode_playback.py
#!/usr/bin/env python3
import sys
import gi
import logging

gi.require_version("GLib", "2.0")
gi.require_version("GObject", "2.0")
gi.require_version("Gst", "1.0")

from gi.repository import Gst, GLib, GObject

logging.basicConfig(level=logging.DEBUG,
                    format="[%(name)s] [%(levelname)8s] - %(message)s")
logger = logging.getLogger(__name__)


#Pad for demux 
def on_pad_added(bus, pad, data):
    srcpad = data
    sinkpad = Gst.Element.get_static_pad(srcpad, "sink")
    Gst.Pad.link(pad, sinkpad)
    return True


if __name__ == "__main__":
    # Initialize GStreamer passing command line argument
    Gst.init(sys.argv)

    loop = GLib.MainLoop()


if len(sys.argv) != 2:
    print("Missing <output-location> parametter")
    sys.exit(1)

# Create the empty pipeline
pipeline = Gst.Pipeline.new("video-player")

# Create the elements
source = Gst.ElementFactory.make("filesrc", "file-source")
sink = Gst.ElementFactory.make("waylandsink", "video-output")
decoder = Gst.ElementFactory.make("avdec_h264", "h264-decoder")
demuxer = Gst.ElementFactory.make("qtdemux", "qt-demuxer")
Videoconv = Gst.ElementFactory.make("videoconvert", "converter")

if not pipeline or not source or not demuxer or not decoder or not sink or not Videoconv:
    logger.error("Element not created.")
    sys.exit(1)
bus = pipeline.get_bus()
bus_id = bus.add_signal_watch()
Gst.Object.unref(bus)

# Build the pipeline
pipeline.add(source)
pipeline.add(demuxer)
pipeline.add(decoder)
pipeline.add(Videoconv)
pipeline.add(sink)


if not source.link(demuxer):
    print("ERROR: Could not link source to demux")
    sys.exit(1)

if not decoder.link(Videoconv):
    print("ERROR: Could not link decoder to video_convert")
    sys.exit(1)

if not Videoconv.link(sink):
    print("ERROR: Could not link scaler to sink")
    sys.exit(1)

source.set_property("location", sys.argv[1])

demuxer.connect("pad-added", on_pad_added, decoder)

try:
    # Start playing
    pipeline.set_state(Gst.State.PLAYING)
    print("Video playback started...")
    print('Press Ctrl+C to stop')
    loop.run()
except KeyboardInterrupt:
    print("Video playback stopped...")
    pipeline.set_state(Gst.State.NULL)
    sys.exit(0)